### ISSUE TITLE
Improved Verification of Tracing Invariants

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#1135]
Legion needs a mechanism for users to verify the invariants required for tracing that is both sound and precise. Today there is only very rudimentary checking and we mostly just trust users to do the right thing. We have a mechanism for verifying that we get the same arguments across shards for control replication and that is by computing hashes of all the argument today for creating each task and operation. We can use the same infrastructure to compute hashes of argument data for creating tasks and operations in traces and then ensure that the hashes are stable across replays. Users should be able to control this checking with a runtime flag the same way that they check control replication invariants with `-lg:safe_ctrlrepl` today.

**Comment by lightsighter** [comment]
Support for this has been added in the master branch with the `-lg:safe_tracing` flag. When enabled, Legion computes a hash for each operation in the trace of all the arguments to that operation and then confirms that the sequence of hashes match under replays of the trace.

**Comment by elliottslaughter** [comment]
Is the presence of `-lg:safe_tracing` reported in the Legion Prof logs? Seems like the sort of thing we should track in case someone enables it and then complains performance isn't good.

**Comment by lightsighter** [comment]
You've actually reminded me to update this code to use the existing hashes that we're already computing for the automatic tracing, which are very cheap (on the order of a couple of microseconds to compute at most) so the cost of turning on this check is nearly free. You could maybe even make the case that it should just be on by default all the time. Either way, I don't think we need to call this out in the profiler as something you need to worry about.

**Comment by elliottslaughter** [comment]
Certainly, if it's this cheap it feels worth turning on by default when DEBUG is enabled?
